src/stun.ts

```diff
@@ -115,67 +143,71 @@ export function getExternalIp(
       tids.splice(index, 1)
       const attr = res.getXorMappedAddressAttribute() ?? res.getMappedAddressAttribute()
 
+      console.log = backup
+
       if (attr == null) {
         error(`STUN response seems to have neither MappedAddress nor XORMappedAddress set. Dropping message`)
-        console.log = backup
         return
       }
 
       verbose(`Received STUN response. External address seems to be: ${attr.address}:${attr.port}`)
 
-      if (results.push(attr) == multiAddrs?.length) {
+      if (results.push(attr) == usableMultiaddrs.length) {
         done()
       }
-
-      console.log = backup
-    }
-
-    const errHandler = (err: any) => {
-      verbose('STUN error', err)
     }
 
     socket.on('message', msgHandler)
-    socket.on('error', errHandler)
 
-    for (const [index, ma] of multiAddrs.entries()) {
-      if (!['ip4', 'ip6', 'dns4', 'dns6'].includes(ma.protoNames()[0])) {
-        error(`Cannot contact STUN server ${ma.toString()} due invalid address.`)
-        return
-      }
+    const allSent = performStunRequests(usableMultiaddrs, tids, socket)
 
-      const nodeAddress = ma.nodeAddress()
+    if (allSent.length > 0) {
+      const sendResults = await Promise.all(allSent)
 
-      const res = stun.createBindingRequest(tids[index]).setFingerprintAttribute()
-
-      verbose(`STUN request sent to ${nodeAddress.address}:${nodeAddress.port}`)
+      if (!sendResults.some((result) => result)) {
```

Comment:
Better variable naming would indicate what is going on here:

```suggestion
      if (!sendResults.some((requestOk: boolean) => requestOk)) {
```

Reply:
changed